packages/next/build/webpack/plugins/flight-types-plugin.ts

```diff
@@ -0,0 +1,133 @@
+  }
+) {
+  return `// File: ${fullPath}
+import * as entry from '${relativePath}'
+type TEntry = typeof entry
+
+check<IEntry, TEntry>(entry)
+
+interface IEntry {
+  ${
+    options.type === 'layout'
+      ? `default: (props: { children: React.ReactNode; params?: any }) => React.ReactNode | null`
+      : `default: (props: { params?: any }) => React.ReactNode | null`
+  }
+  generateStaticParams?: (params?:any) => Promise<any[]>
+  config?: {
+    // TODO: remove revalidate here
+    revalidate?: number | boolean
+    ${options.type === 'page' ? 'runtime?: string' : ''}
+  }
+  revalidate?: RevalidateRange<TEntry> | false
+  dynamic?: 'auto' | 'force-dynamic' | 'error' | 'force-static'
+  dynamicParams?: boolean
+  fetchCache?: 'auto' | 'force-no-store' | 'only-no-store' | 'default-no-store' | 'default-cache' | 'only-cache' | 'force-cache'
+  preferredRegion?: 'auto' | 'home' | 'edge'
+}
+
+// =============
+// Utility types
+type RevalidateRange<T> = T extends { revalidate: any } ? NonNegative<T['revalidate']> : never
+type Impossible<K extends keyof any> = { [P in K]: never }
+function check<Base, T extends Base>(_mod: T & Impossible<Exclude<keyof T, keyof Base>>): void {}
+
+// https://github.com/sindresorhus/type-fest
+type Numeric = number | bigint
+type Zero = 0 | 0n
+type Negative<T extends Numeric> = T extends Zero ? never : \`\${T}\` extends \`-\${string}\` ? T : never
+type NonNegative<T extends Numeric> = T extends Zero ? T : Negative<T> extends never ? T : '__invalid_negative_number__'
+`
+}
+
+export class FlightTypesPlugin {
+  appDir: string
+  dev: boolean
+  isEdgeServer: boolean
+
+  constructor(options: Options) {
+    this.appDir = options.appDir
+    this.dev = options.dev
+    this.isEdgeServer = options.isEdgeServer
+  }
+
+  apply(compiler: webpack.Compiler) {
+    const assetPrefix = this.dev
+      ? '..'
+      : this.isEdgeServer
+      ? '..'
+      : path.join('..', '..')
+
+    const handleModule = (_mod: webpack.Module, assets: any) => {
+      if (_mod.layer !== WEBPACK_LAYERS.server) return
+      const mod: webpack.NormalModule = _mod as any
+
+      if (!mod.resource) return
+      if (!mod.resource.startsWith(this.appDir + path.sep)) return
+      if (!/\.(js|jsx|ts|tsx|mjs)$/.test(mod.resource)) return
+
+      const IS_LAYOUT = /[/\\]layout\.[^.]+$/.test(mod.resource)
+      const IS_PAGE = !IS_LAYOUT && /[/\\]page\.[^.]+$/.test(mod.resource)
+      const relativePath = path.relative(this.appDir, mod.resource)
+
+      // const RSC = mod.buildInfo.rsc
+
+      const typePath = path.join(
+        'types',
+        'app',
+        relativePath.replace(/\.(js|jsx|ts|tsx|mjs)$/, '.ts')
+      )
+      const relativeImportPath = path.join(
+        path.relative(typePath, ''),
+        'app',
+        relativePath.replace(/\.(js|jsx|ts|tsx|mjs)$/, '')
+      )
+      const assetPath = path.join(assetPrefix, typePath)
+
+      if (IS_LAYOUT) {
+        assets[assetPath] = new sources.RawSource(
```

Comment:
This is a fresh Source instance on every compilation, so it will try to write all of them on each recompilation. Use a cached instance.

Reply:
But that doesn't really matter since the plugin is only applied on production builds. Shouldn't page type be tested in dev too? I guess here it would matter the most?